lib/src/domain/repository/repository_card.dart

```diff
@@ -1,28 +1,109 @@
+  late List<HearthstoneCard> cardListApi = <HearthstoneCard>[];
 
   RepositoryCard({
     required this.client,
+    required this.cardDatabase,
     required this.cardsApi,
   });
 
-  late List<HearthstoneCard> cardListApi = <HearthstoneCard>[];
+  Future<List<dynamic>> fetchDocumentSnapshot({
+    required String mainCollectionDoc,
+    required String subCollection,
+  }) async {
+    final List<dynamic> documentList = [];
+    final QuerySnapshot documentSnapshot = await cardDatabase.readItems(
+      mainCollectionDoc: mainCollectionDoc,
+      subCollectionDoc: subCollection,
+    );
+
+    final List<QueryDocumentSnapshot> docCards = documentSnapshot.docs;
+
+    for (final doc in docCards) {
+      documentList.add(doc.data());
+    }
+    return documentList;
+  }
+
+  void addCardBase(
+    String endpoint,
+    List<dynamic> cardsApiList,
+  ) {
+    switch (endpoint) {
+      case kRacesDemonEndPointApi:
+        cardDatabase.addItem(
+          cardList: cardsApiList,
+          mainCollectionDoc: kMainCollectionDocDemon,
+          subCollectionDoc: kSubCollectionDemon,
+        );
+        break;
+      case kRacesTotemEndPointApi:
+        cardDatabase.addItem(
+          cardList: cardsApiList,
+          mainCollectionDoc: kMainCollectionDocTotem,
+          subCollectionDoc: kSubCollectionTotem,
+        );
+        break;
+      case kRacesMurlocEndPointApi:
+        cardDatabase.addItem(
+          cardList: cardsApiList,
+          mainCollectionDoc: kMainCollectionDocMurloc,
+          subCollectionDoc: kSubCollectionMurloc,
+        );
+        break;
+    }
+  }
 
   Future<List<HearthstoneCard>> addCardList(String endpoint) async {
     cardListApi = [];
-    final String responseBody = await cardsApi.fetchCardList(
-      endpoint,
-      client,
-    );
-    final List<dynamic> cardsApiList = await jsonDecode(
-      responseBody,
-    ) as List<dynamic>;
+    List<dynamic> cardsApiList = [];
+    try {
+      final responseBody = await cardsApi.fetchCardList(
+        endpoint,
+        client,
+      );
+      cardsApiList = await jsonDecode(
+        responseBody,
+      ) as List<dynamic>;
+      addCardBase(
+        endpoint,
+        cardsApiList,
+      );
+    } catch (database) {
+      switch (endpoint) {
+        case kRacesDemonEndPointApi:
+          cardsApiList = await fetchDocumentSnapshot(
+            mainCollectionDoc: kMainCollectionDocDemon,
+            subCollection: kSubCollectionDemon,
+          );
+          break;
+        case kRacesTotemEndPointApi:
+          cardsApiList = await fetchDocumentSnapshot(
+            mainCollectionDoc: kMainCollectionDocTotem,
+            subCollection: kSubCollectionTotem,
+          );
+          break;
+        case kRacesMurlocEndPointApi:
+          cardsApiList = await fetchDocumentSnapshot(
+            mainCollectionDoc: kMainCollectionDocMurloc,
+            subCollection: kSubCollectionMurloc,
+          );
+          break;
+      }
+    }
```

Comment:
what happens when the database is empty and the service fails? how do you catch the exception?

Reply:
It returns an empty list [].

Reply:
Done.